packages/csv-parser-state/src/main.js

```diff
@@ -0,0 +1,190 @@
+import type {
+  ApiConfigOptions,
+  ConstructorOptions,
+  CsvOptions,
+  LoggerOptions,
+  StateWithStringTransitions,
+  StateWithUnresolvedTransitions,
+} from 'types/state'
+import type { Client, SuccessResult } from 'types/sdk'
+import pkg from '../package.json'
+
+export default class CsvParserState {
+  // Set flowtype annotations
+  accessToken: string
+  apiConfig: ApiConfigOptions
+  client: Client
+  continueOnProblems: boolean
+  csvConfig: CsvOptions
+  logger: LoggerOptions
+  _rowIndex: number
+  _fetchStates: Function
+
+  constructor(
+    {
+      apiConfig,
+      csvConfig,
+      continueOnProblems = false,
+      accessToken,
+    }: ConstructorOptions,
+    logger: LoggerOptions
+  ) {
+    this.apiConfig = apiConfig
+    this.accessToken = accessToken
+
+    const defaultOptions = {
+      delimiter: ',',
+      multiValueDelimiter: ';',
+    }
+    this.csvConfig = { ...defaultOptions, ...csvConfig }
+    this.continueOnProblems = continueOnProblems
+    this.logger = logger || {
+      error: () => {},
+      warn: () => {},
+      info: () => {},
+      debug: () => {},
+    }
+
+    this._rowIndex = 0
+  }
+
+  parse(input: stream$Readable, output: stream$Writable) {
+    this.logger.info('Starting conversion')
+    highland(input)
+      .through(csv({ separator: this.csvConfig.delimiter, strict: true }))
+      .map(CsvParserState._removeEmptyFields)
+      .map(CsvParserState._parseInitialToBoolean)
+      .map(unflatten)
+      .map(state => this._mapMultiValueFieldsToArray(state))
+      .flatMap(state => highland(this._transformTransitions(state)))
+      .errors((error, cb) => this._handleErrors(error, cb))
+      .stopOnError(error => {
+        // <- Emit error and close stream if needed
+        this.logger.error(`At row: ${this._rowIndex}, ${error}`)
+        output.emit('error', error)
+      })
+      .doto(() => {
+        this._rowIndex += 1
+        this.logger.debug(`At row: ${this._rowIndex}, Successfully parsed`)
+      })
+      .pipe(JSONStream.stringify())
+      .pipe(output)
+  }
+
+  async _transformTransitions({
+    transitions,
+    ...restOfState
+  }: StateWithUnresolvedTransitions) {
+    if (!transitions) return restOfState
+    // We setup the client here because it is unnecessary
+    // if there are no transitions
+    this._setupClient()
+    const stateRequests = transitions.map(transitionState =>
+      this._buildStateRequest(transitionState)
+    )
+    const resolvedStates = await Promise.all(stateRequests)
+    const stateReferences = resolvedStates.map(response => ({
+      typeId: 'state',
+      id: response.body.id,
+    }))
+    return { ...restOfState, transitions: stateReferences }
+  }
+
+  _buildStateRequest(stateKey: string): Promise<SuccessResult> {
+    const stateService = this._createStateService()
+    const uri = stateService.byKey(stateKey).build()
+    return this._fetchStates(uri)
+  }
+
+  _setupClient() {
```

Comment:
Nitpick: createCliebt?

Reply:
I'm wondering why this is part of the instance in the first place.
The only time the `client` itself is used is with `execute` below.